src/magic_folder/config.py

```diff
@@ -0,0 +1,419 @@
+                        dbversion,
+                    )
+                )
+
+    @property
+    def author(self):
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute("SELECT author_name, author_private_key FROM config");
+            name, keydata = cursor.fetchone()
+            return LocalAuthor(
+                name=name,
+                signing_key=SigningKey(keydata, encoder=Base32Encoder),
+            )
+
+    @property
+    def stash_path(self):
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute("SELECT stash_path FROM config");
+            path_raw = cursor.fetchone()
+            return FilePath(path_raw)
+
+
+@attr.s
+class GlobalConfigDatabase(object):
+    """
+    Low-level access to the global configuration database
+    """
+    database = attr.ib()  # sqlite3 Connection; needs validator
+    api_token_path = attr.ib(validator=attr.validators.instance_of(FilePath))
+
+    def __attrs_post_init__(self):
+        self._api_token = None
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute("BEGIN IMMEDIATE TRANSACTION")
+            cursor.execute("SELECT version FROM version");
+            dbversion = cursor.fetchone()[0]
+            if dbversion != _global_config_version:
+                raise ConfigurationError(
+                    "Unknown configuration database version (wanted {}, got {})".format(
+                        _global_config_version,
+                        dbversion,
+                    )
+                )
+
+    @property
+    def api_token(self):
+        """
+        Current API token
+        """
+        if self._api_token is None:
+            with self.api_token_path.open('rb') as f:
+                self._api_token = f.read()
+        return self._api_token
+
+    def rotate_api_token(self):
+        """
+        Record a new random API token and then return it
+        """
+        # this goes directly into Web headers, so we use the same
+        # encoding as Tahoe uses.
+        self._api_token = urlsafe_b64encode(urandom(32))
+        with self.api_token_path.open('wb') as f:
+            f.write(self._api_token)
+        return self._api_token
+
+    @property
+    def api_endpoint(self):
+        """
+        The twisted server-string describing our API listener
+        """
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute("SELECT api_endpoint FROM config")
+            return cursor.fetchone()[0]
+
+    @api_endpoint.setter
+    def api_endpoint(self, ep_string):
+        # confirm we have a valid endpoint-string
+        from twisted.internet import reactor  # uhm...
+        # XXX so, having the reactor here sucks. But if we pass in an
+        # IStreamServerEndpoint instead, how can we turn that back
+        # into an endpoint-string?
+        serverFromString(reactor, ep_string)
```

Comment:
Yea, all that stuff you wrote.  There is no going from an IStreamServerEndpoint to a string, though.  The interface is not expressive enough.

And the endpoint parser is private so you can't really validate the endpoint in any way except with `serverFromString`.  That is pretty lame.  Maybe most of them don't use the reactor for parsing/initialization and we could get away with passing `None`? :/  Still pretty lame.